omniledger/service/struct.go

```diff
@@ -175,3 +179,94 @@ type Data struct {
 	Timestamp    int64
 	Roster       *onet.Roster
 }
+
+// sortTransactions needs to marshal transactions, if it fails to do so,
+// it returns an error and leaves the slice unchange.
+// The helper functions (sortWithSalt, xorTransactions) operate on []byte
+// representations directly. This allows for some more compact error handling
+// when (un)marshalling.
+func sortTransactions(ts []Transaction) error {
+	bs := make([][]byte, len(ts))
+	sortedTs := make([]*Transaction, len(ts))
+	var err error
+	var ok bool
+	for i := range ts {
+		bs[i], err = network.Marshal(&ts[i])
+		if err != nil {
+			return err
+		}
+	}
+	// An alternative to XOR-ing the transactions would have been to
+	// concatenate them and hash the result. However, if we generate the salt
+	// as the hash of the concatenation of the transactions, we have to
+	// concatenate them in a specific order to be deterministic.
+	// This means we would have to sort them, just to get the salt.
+	// In order to avoid this, we XOR them.
+	salt := xorTransactions(bs)
+	sortWithSalt(bs, salt)
+	for i := range bs {
+		_, tmp, err := network.Unmarshal(bs[i], cothority.Suite)
+		if err != nil {
+			return err
+		}
+		sortedTs[i], ok = tmp.(*Transaction)
+		if !ok {
+			return errors.New("Data of wrong type")
+		}
+	}
+	for i := range sortedTs {
+		ts[i] = *sortedTs[i]
+	}
+	return nil
+}
+
+// sortWithSalt sorts transactions according to their salted hash:
+// The salt is prepended to the transactions []byte representation
+// and this concatenation is hashed then.
+// Using a salt here makes the resulting order of the transactions
+// harder to guess.
+func sortWithSalt(ts [][]byte, salt []byte) {
+	less := func(i, j int) bool {
+		h := sha256.New()
+		h.Write(append(salt, ts[i]...))
+		h1 := h.Sum(nil)
+		h.Reset()
+		h.Write(append(salt, ts[j]...))
+		h2 := h.Sum(nil)
+		h.Reset()
+		result := bytes.Compare(h1, h2) == -1
+		return result
+	}
+	sort.Slice(ts, less)
+}
+
+// xorTransactions returns the XOR of the hash values of all the transactions.
+func xorTransactions(ts [][]byte) []byte {
+	h := sha256.New()
+	h.Write(nil)
+	hn := h.Sum(nil)
+	h.Reset()
+	result := make([]byte, len(hn))
```

Comment:
hn appears to be just used for computing the length, I think it's cleaner to just use https://golang.org/pkg/crypto/sha256/#pkg-constants

Reply:
And then remove `safeXORBytes`, as it is not needed. Just `for` loop over the bytes and xor them.